lib/Solver/Solver.cpp

```diff
@@ -0,0 +1,363 @@
+#include "vara/Solver/Solver.h"
+
+#include "z3++.h"
+
+namespace vara::solver {
+
+Result<SolverErrorCode>
+Z3Solver::addFeature(const feature::Feature &FeatureToAdd,
+                     bool IsInAlternativeGroup) {
+  // Check whether the parent feature is already added
+  vara::feature::Feature *Parent = FeatureToAdd.getParentFeature();
+  if (Parent != nullptr && OptionToVariableMapping.find(Parent->getName()) ==
+                               OptionToVariableMapping.end()) {
+    return PARENT_NOT_PRESENT;
+  }
+
+  if (OptionToVariableMapping.find(FeatureToAdd.getName()) !=
+      OptionToVariableMapping.end()) {
+    return ALREADY_PRESENT;
+  }
+
+  // Add the feature
+  switch (FeatureToAdd.getKind()) {
+  case feature::Feature::FeatureKind::FK_NUMERIC: {
+    const auto *F =
+        llvm::dyn_cast<vara::feature::NumericFeature>(&FeatureToAdd);
+    if (!F) {
+      return NOT_SUPPORTED;
+    }
+    const auto Values = F->getValues();
+    if (std::holds_alternative<vara::feature::NumericFeature::ValueListType>(
+            Values)) {
+      if (auto R = addFeature(
+              F->getName().str(),
+              std::get<vara::feature::NumericFeature::ValueListType>(Values));
+          !R) {
+        return R;
+      }
+    } else {
+      auto Range =
+          std::get<vara::feature::NumericFeature::ValueRangeType>(Values);
+      auto *StepFunction = F->getStepFunction();
+      auto Step = Range.first;
+      std::vector<int64_t> Vals;
+      while (Step < Range.second) {
+        Vals.insert(Vals.begin(), Step);
+        Step = StepFunction->next(Step);
+      }
+      if (auto R = addFeature(F->getName().str(), Vals); !R) {
+        return R;
+      }
+    }
+    break;
+  }
+  case feature::Feature::FeatureKind::FK_BINARY: {
+    addFeature(FeatureToAdd.getName().str());
+    // Add all constraints (i.e., implications, exclusions, parent feature)
+    if (!llvm::dyn_cast<vara::feature::BinaryFeature>(&FeatureToAdd)) {
```

Comment:
```suggestion
    if (!llvm::isa<vara::feature::BinaryFeature>(&FeatureToAdd)) {
```
I would also move this above the `addFeature` call an check as early as possible